File: frontend/src/lib/engine/macro.ts
import { createServerFn } from "@tanstack/react-start";

import { fetchBinanceKlinesDirect } from "./binance";

export type MacroInstrumentId = "spx" | "ndx" | "dxy" | "gold";
export type CorrelationRegime = "coupled" | "decoupled" | "inverse";

export interface MacroInstrument {
  id: MacroInstrumentId;
  label: string;
  last: number;
  changePercent: number;
  spark: number[];
}

export interface MacroSnapshot {
  instruments: MacroInstrument[];
  /** Pearson correlation of BTC vs Nasdaq 100 daily returns, ~30 shared sessions. */
  btcNdxCorrelation: number | null;
  correlationRegime: CorrelationRegime | null;
  source: "live" | "demo";
  updatedAt: string;
}

interface SeriesPoint {
  date: string;
  close: number;
}

const INSTRUMENT_LABELS: Record<MacroInstrumentId, string> = {
  spx: "S&P 500",
  ndx: "Nasdaq 100",
  dxy: "Dollar Index",
  gold: "Gold",
};

// FRED publishes these daily with a short lag and no API key; Yahoo/Stooq
// block datacenter IPs. Gold rides the Binance pipeline via PAXG (tokenized
// gold, tracks spot ~1:1) so it stays live even when FRED lags.
const FRED_SERIES: Record<Exclude<MacroInstrumentId, "gold">, string> = {
  spx: "SP500",
  ndx: "NASDAQ100",
  dxy: "DTWEXBGS",
};

const SPARK_POINTS = 30;
const CORRELATION_SESSIONS = 30;

async function fetchFredSeries(id: string): Promise<SeriesPoint[]> {
  const start = new Date(Date.now() - 150 * 24 * 60 * 60 * 1000).toISOString().slice(0, 10);
  try {
    const response = await fetch(
      `https://fred.stlouisfed.org/graph/fredgraph.csv?id=${id}&cosd=${start}`,
    );
    if (!response.ok) return [];
    const text = await response.text();
    return text
      .split("\n")
      .slice(1)
      .flatMap((line) => {
        const [date, raw] = line.trim().split(",");
        const close = Number.parseFloat(raw);
        return date && Number.isFinite(close) ? [{ date, close }] : [];
      });
  } catch {
    return [];
  }
}

async function fetchBinanceDailySeries(symbol: string): Promise<SeriesPoint[]> {
  const candles = await fetchBinanceKlinesDirect({ symbol, timeframe: "1D", limit: 120 });
  return candles.map((c) => ({
    date: new Date(c.time * 1000).toISOString().slice(0, 10),
    close: c.close,
  }));
}

function toInstrument(id: MacroInstrumentId, series: SeriesPoint[]): MacroInstrument {
  const closes = series.map((p) => p.close);
  const last = closes.at(-1) ?? 0;
  const prev = closes.at(-2) ?? last;
  return {
    id,
    label: INSTRUMENT_LABELS[id],
    last: Number(last.toFixed(2)),
    changePercent: prev !== 0 ? Number((((last - prev) / prev) * 100).toFixed(2)) : 0,
    spark: closes.slice(-SPARK_POINTS).map((v) => Number(v.toFixed(2))),
  };
}

function pearson(a: number[], b: number[]): number | null {
  const n = Math.min(a.length, b.length);
  if (n < 10) return null;
  const meanA = a.reduce((s, v) => s + v, 0) / n;
  const meanB = b.reduce((s, v) => s + v, 0) / n;
  let cov = 0;
  let varA = 0;
  let varB = 0;
  for (let i = 0; i < n; i++) {
    const da = a[i] - meanA;
    const db = b[i] - meanB;
    cov += da * db;
    varA += da * da;
    varB += db * db;
  }
  if (varA === 0 || varB === 0) return null;
  return cov / Math.sqrt(varA * varB);
}

/**
 * Correlation of daily returns over the last ~30 sessions both markets traded.
 * Aligning on shared dates drops crypto's weekends so a closed stock market
 * never reads as divergence.
 */
export function computeBtcNdxCorrelation(btc: SeriesPoint[], ndx: SeriesPoint[]): number | null {
  const btcByDate = new Map(btc.map((p) => [p.date, p.close]));
  const shared = ndx.filter((p) => btcByDate.has(p.date)).slice(-(CORRELATION_SESSIONS + 1));
  if (shared.length < 11) return null;

  const btcReturns: number[] = [];
  const ndxReturns: number[] = [];
  for (let i = 1; i < shared.length; i++) {
    const prevBtc = btcByDate.get(shared[i - 1].date) as number;
    const curBtc = btcByDate.get(shared[i].date) as number;
    if (prevBtc === 0 || shared[i - 1].close === 0) continue;
    btcReturns.push(curBtc / prevBtc - 1);
    ndxReturns.push(shared[i].close / shared[i - 1].close - 1);
  }
  const value = pearson(btcReturns, ndxReturns);
  return value === null ? null : Number(value.toFixed(2));
}

export function correlationRegimeOf(correlation: number | null): CorrelationRegime | null {
  if (correlation === null) return null;
  if (correlation >= 0.4) return "coupled";
  if (correlation <= -0.3) return "inverse";
  return "decoupled";
}

// --- deterministic demo fallback (same philosophy as mock-candles) ---

const DEMO_BASE: Record<MacroInstrumentId, { base: number; volatility: number }> = {
  spx: { base: 7480, volatility: 0.008 },
  ndx: { base: 29500, volatility: 0.012 },
  dxy: { base: 121, volatility: 0.003 },
  gold: { base: 4170, volatility: 0.009 },
};

function demoSeries(id: string, base: number, volatility: number): SeriesPoint[] {
  let seed = 0;
  for (let i = 0; i < id.length; i++) seed = (seed * 31 + id.charCodeAt(i)) >>> 0;
  const rand = () => {
    seed = (seed * 1664525 + 1013904223) >>> 0;
    return seed / 4294967296;
  };
  const end = Date.UTC(2026, 6, 5);
  let close = base * (0.97 + rand() * 0.06);
  return Array.from({ length: 60 }, (_, i) => {
    close *= 1 + (rand() - 0.49) * volatility;
    return {
      date: new Date(end - (59 - i) * 24 * 60 * 60 * 1000).toISOString().slice(0, 10),
      close,
    };
  });
}

export function buildDemoMacroSnapshot(): MacroSnapshot {
  const series = Object.fromEntries(
    (Object.keys(DEMO_BASE) as MacroInstrumentId[]).map((id) => [
      id,
      demoSeries(id, DEMO_BASE[id].base, DEMO_BASE[id].volatility),
    ]),
  ) as Record<MacroInstrumentId, SeriesPoint[]>;
  const btcDemo = demoSeries("btc", 108_000, 0.02);
  const correlation = computeBtcNdxCorrelation(btcDemo, series.ndx);
  return {
    instruments: (Object.keys(DEMO_BASE) as MacroInstrumentId[]).map((id) =>
      toInstrument(id, series[id]),
    ),
    btcNdxCorrelation: correlation,
    correlationRegime: correlationRegimeOf(correlation),
    source: "demo",
    updatedAt: new Date().toISOString(),
  };
}

let macroCache: { at: number; data: MacroSnapshot } | null = null;
const MACRO_TTL_MS = 10 * 60_000;

async function computeMacroSnapshot(): Promise<MacroSnapshot> {
  const now = Date.now();
  if (macroCache && now - macroCache.at < MACRO_TTL_MS) return macroCache.data;

  const [spx, ndx, dxy, gold, btc] = await Promise.all([
    fetchFredSeries(FRED_SERIES.spx),
    fetchFredSeries(FRED_SERIES.ndx),
    fetchFredSeries(FRED_SERIES.dxy),
    fetchBinanceDailySeries("PAXG"),
    fetchBinanceDailySeries("BTC"),
  ]);

  const series: Record<MacroInstrumentId, SeriesPoint[]> = { spx, ndx, dxy, gold };
  const liveCount = Object.values(series).filter((s) => s.length >= 20).length;
  if (liveCount < 3) {
    const demo = buildDemoMacroSnapshot();
    macroCache = { at: now, data: demo };
    return demo;
  }

  const correlation =
    btc.length >= 20 && ndx.length >= 20 ? computeBtcNdxCorrelation(btc, ndx) : null;
  const snapshot: MacroSnapshot = {
    instruments: (Object.keys(INSTRUMENT_LABELS) as MacroInstrumentId[])
      .filter((id) => series[id].length >= 2)
      .map((id) => toInstrument(id, series[id])),
    btcNdxCorrelation: correlation,
    correlationRegime: correlationRegimeOf(correlation),
    source: "live",
    updatedAt: new Date().toISOString(),
  };
  macroCache = { at: now, data: snapshot };
  return snapshot;
}

export const fetchMacroSnapshotServer = createServerFn({ method: "GET" }).handler(async () =>
  computeMacroSnapshot(),
);

/** Client-facing helper: server snapshot, falling back to the deterministic demo build. */
export async function fetchMacroSnapshot(): Promise<MacroSnapshot> {
  try {
    return await fetchMacroSnapshotServer();
  } catch {
    return buildDemoMacroSnapshot();
  }
}

File: engine/smc/macro.py
"""Macro (TradFi) context read models (port of macro.ts, pure parts).

S&P 500 / Nasdaq 100 / Dollar Index / Gold instruments plus the BTC↔NDX
correlation regime. The TS module's FRED/Binance fetchers and cache are
backend concerns; the engine turns daily close series it is given into
instruments and correlation reads. `FRED_SERIES` stays here as shared config
so every fetcher asks for the same series.
"""

import math
from dataclasses import dataclass
from datetime import UTC, date, datetime, timedelta
from typing import Literal

MacroInstrumentId = Literal["spx", "ndx", "dxy", "gold"]
CorrelationRegime = Literal["coupled", "decoupled", "inverse"]

MACRO_INSTRUMENT_IDS: tuple[MacroInstrumentId, ...] = ("spx", "ndx", "dxy", "gold")


@dataclass(slots=True)
class MacroInstrument:
    id: MacroInstrumentId
    label: str
    last: float
    change_percent: float
    spark: list[float]


@dataclass(slots=True)
class MacroSnapshot:
    instruments: list[MacroInstrument]
    # Pearson correlation of BTC vs Nasdaq 100 daily returns, ~30 shared sessions.
    btc_ndx_correlation: float | None
    correlation_regime: CorrelationRegime | None
    source: Literal["live", "demo"]
    updated_at: str


@dataclass(slots=True)
class SeriesPoint:
    date: str
    close: float


INSTRUMENT_LABELS: dict[MacroInstrumentId, str] = {
    "spx": "S&P 500",
    "ndx": "Nasdaq 100",
    "dxy": "Dollar Index",
    "gold": "Gold",
}

# FRED publishes these daily with a short lag and no API key; Yahoo/Stooq block
# datacenter IPs. Gold rides the Binance pipeline via PAXG (tokenized gold,
# tracks spot ~1:1) so it stays live even when FRED lags.
FRED_SERIES: dict[str, str] = {
    "spx": "SP500",
    "ndx": "NASDAQ100",
    "dxy": "DTWEXBGS",
}

SPARK_POINTS = 30
CORRELATION_SESSIONS = 30


def _to_fixed2(value: float) -> float:
    """Number(x.toFixed(2)) — half toward +infinity is close enough off the decision path."""
    return math.floor(value * 100 + 0.5) / 100


def to_instrument(id_: MacroInstrumentId, series: list[SeriesPoint]) -> MacroInstrument:
    closes = [p.close for p in series]
    last = closes[-1] if closes else 0.0
    prev = closes[-2] if len(closes) >= 2 else last
    return MacroInstrument(
        id=id_,
        label=INSTRUMENT_LABELS[id_],
        last=_to_fixed2(last),
        change_percent=_to_fixed2((last - prev) / prev * 100) if prev != 0 else 0.0,
        spark=[_to_fixed2(v) for v in closes[-SPARK_POINTS:]],
    )


def _pearson(a: list[float], b: list[float]) -> float | None:
    n = min(len(a), len(b))
    if n < 10:
        return None
    mean_a = sum(a[:n]) / n
    mean_b = sum(b[:n]) / n
    cov = 0.0
    var_a = 0.0
    var_b = 0.0
    for x, y in zip(a[:n], b[:n], strict=True):
        da = x - mean_a
        db = y - mean_b
        cov += da * db
        var_a += da * da
        var_b += db * db
    if var_a == 0 or var_b == 0:
        return None
    return cov / math.sqrt(var_a * var_b)


def compute_btc_ndx_correlation(btc: list[SeriesPoint], ndx: list[SeriesPoint]) -> float | None:
    """Correlation of daily returns over the last ~30 sessions both markets traded.

    Aligning on shared dates drops crypto's weekends so a closed stock market
    never reads as divergence.
    """
    btc_by_date = {p.date: p.close for p in btc}
    shared = [p for p in ndx if p.date in btc_by_date][-(CORRELATION_SESSIONS + 1) :]
    if len(shared) < 11:
        return None

    btc_returns: list[float] = []
    ndx_returns: list[float] = []
    for i in range(1, len(shared)):
        prev_btc = btc_by_date[shared[i - 1].date]
        cur_btc = btc_by_date[shared[i].date]
        if prev_btc == 0 or shared[i - 1].close == 0:
            continue
        btc_returns.append(cur_btc / prev_btc - 1)
        ndx_returns.append(shared[i].close / shared[i - 1].close - 1)
    value = _pearson(btc_returns, ndx_returns)
    return None if value is None else _to_fixed2(value)


def correlation_regime_of(correlation: float | None) -> CorrelationRegime | None:
    if correlation is None:
        return None
    if correlation >= 0.4:
        return "coupled"
    if correlation <= -0.3:
        return "inverse"
    return "decoupled"


# --- deterministic demo fallback (same philosophy as mock_candles) ---

_DEMO_BASE: dict[MacroInstrumentId, tuple[float, float]] = {
    "spx": (7480, 0.008),
    "ndx": (29500, 0.012),
    "dxy": (121, 0.003),
    "gold": (4170, 0.009),
}

_DEMO_END = date(2026, 7, 5)


def _demo_series(id_: str, base: float, volatility: float) -> list[SeriesPoint]:
    seed = 0
    for ch in id_:
        seed = (seed * 31 + ord(ch)) & 0xFFFFFFFF

    def rand() -> float:
        nonlocal seed
        seed = (seed * 1664525 + 1013904223) & 0xFFFFFFFF
        return seed / 4294967296

    close = base * (0.97 + rand() * 0.06)
    out: list[SeriesPoint] = []
    for i in range(60):
        close *= 1 + (rand() - 0.49) * volatility
        out.append(SeriesPoint(date=(_DEMO_END - timedelta(days=59 - i)).isoformat(), close=close))
    return out


def build_demo_macro_snapshot() -> MacroSnapshot:
    series = {
        id_: _demo_series(id_, _DEMO_BASE[id_][0], _DEMO_BASE[id_][1])
        for id_ in MACRO_INSTRUMENT_IDS
    }
    btc_demo = _demo_series("btc", 108_000, 0.02)
    correlation = compute_btc_ndx_correlation(btc_demo, series["ndx"])
    return MacroSnapshot(
        instruments=[to_instrument(id_, series[id_]) for id_ in MACRO_INSTRUMENT_IDS],
        btc_ndx_correlation=correlation,
        correlation_regime=correlation_regime_of(correlation),
        source="demo",
        updated_at=datetime.now(UTC).isoformat(),
    )
